evm/src/bn254.rs

```diff
@@ -0,0 +1,743 @@
+    (BN_BASE + x - y) % BN_BASE
+}
+
+fn neg_fp(x: Fp) -> Fp {
+    (BN_BASE - x) % BN_BASE
+}
+
+fn exp_fp(x: Fp, e: U256) -> Fp {
+    let mut current = x;
+    let mut product = U256::one();
+
+    for j in 0..256 {
+        if e.bit(j) {
+            product = U256::try_from(product.full_mul(current) % BN_BASE).unwrap();
+        }
+        current = U256::try_from(current.full_mul(current) % BN_BASE).unwrap();
+    }
+    product
+}
+
+fn inv_fp(x: Fp) -> Fp {
+    exp_fp(x, BN_BASE - 2)
+}
+
+fn div_fp(x: Fp, y: Fp) -> Fp {
+    mul_fp(x, inv_fp(y))
+}
+
+fn conj_fp2(a: Fp2) -> Fp2 {
+    let [a, a_] = a;
+    [a, neg_fp(a_)]
+}
+
+fn add_fp2(a: Fp2, b: Fp2) -> Fp2 {
+    let [a, a_] = a;
+    let [b, b_] = b;
+    [add_fp(a, b), add_fp(a_, b_)]
+}
+
+fn add3_fp2(a: Fp2, b: Fp2, c: Fp2) -> Fp2 {
+    let [a, a_] = a;
+    let [b, b_] = b;
+    let [c, c_] = c;
+    [add3_fp(a, b, c), add3_fp(a_, b_, c_)]
+}
+
+fn sub_fp2(a: Fp2, b: Fp2) -> Fp2 {
+    let [a, a_] = a;
+    let [b, b_] = b;
+    [sub_fp(a, b), sub_fp(a_, b_)]
+}
+
+fn mul_fp2(a: Fp2, b: Fp2) -> Fp2 {
+    let [a, a_] = a;
+    let [b, b_] = b;
+    [
+        sub_fp(mul_fp(a, b), mul_fp(a_, b_)),
+        add_fp(mul_fp(a, b_), mul_fp(a_, b)),
+    ]
+}
+
+fn i9(a: Fp2) -> Fp2 {
+    let [a, a_] = a;
+    let nine = U256::from(9);
+    [sub_fp(mul_fp(nine, a), a_), add_fp(a, mul_fp(nine, a_))]
+}
+
+fn add_fp6(c: Fp6, d: Fp6) -> Fp6 {
+    let [c0, c1, c2] = c;
+    let [d0, d1, d2] = d;
+
+    let e0 = add_fp2(c0, d0);
+    let e1 = add_fp2(c1, d1);
+    let e2 = add_fp2(c2, d2);
+    [e0, e1, e2]
+}
+
+fn sub_fp6(c: Fp6, d: Fp6) -> Fp6 {
+    let [c0, c1, c2] = c;
+    let [d0, d1, d2] = d;
+
+    let e0 = sub_fp2(c0, d0);
+    let e1 = sub_fp2(c1, d1);
+    let e2 = sub_fp2(c2, d2);
+    [e0, e1, e2]
+}
+
+fn neg_fp6(a: Fp6) -> Fp6 {
+    sub_fp6(embed_fp6(ZERO), a)
+}
+
+fn mul_fp6(c: Fp6, d: Fp6) -> Fp6 {
+    let [c0, c1, c2] = c;
+    let [d0, d1, d2] = d;
+
+    let c0d0 = mul_fp2(c0, d0);
+    let c0d1 = mul_fp2(c0, d1);
+    let c0d2 = mul_fp2(c0, d2);
+    let c1d0 = mul_fp2(c1, d0);
+    let c1d1 = mul_fp2(c1, d1);
+    let c1d2 = mul_fp2(c1, d2);
+    let c2d0 = mul_fp2(c2, d0);
+    let c2d1 = mul_fp2(c2, d1);
+    let c2d2 = mul_fp2(c2, d2);
+    let cd12 = add_fp2(c1d2, c2d1);
+
+    [
+        add_fp2(c0d0, i9(cd12)),
+        add3_fp2(c0d1, c1d0, i9(c2d2)),
+        add3_fp2(c0d2, c1d1, c2d0),
+    ]
+}
+
+fn sh(c: Fp6) -> Fp6 {
+    let [c0, c1, c2] = c;
+    [i9(c2), c0, c1]
+}
+
+fn sparse_embed(g0: Fp, g1: Fp2, g2: Fp2) -> Fp12 {
+    [
+        [embed_fp2(g0), g1, embed_fp2(ZERO)],
+        [embed_fp2(ZERO), g2, embed_fp2(ZERO)],
+    ]
+}
+
+pub fn mul_fp12(f: Fp12, g: Fp12) -> Fp12 {
+    let [f0, f1] = f;
+    let [g0, g1] = g;
+
+    let h0 = mul_fp6(f0, g0);
+    let h1 = mul_fp6(f1, g1);
+    let h01 = mul_fp6(add_fp6(f0, f1), add_fp6(g0, g1));
+    [add_fp6(h0, sh(h1)), sub_fp6(h01, add_fp6(h0, h1))]
+}
+
+fn frob_fp6(n: usize, c: Fp6) -> Fp6 {
```

Comment:
The words "Frobenius endomorphism" should be in a comment somewhere, along with descriptions of the arguments and a brief explanation of the algorithm (or a link to an explanation).